fars/booking/gcal.py

```diff
@@ -0,0 +1,215 @@
+                    },
+                },
+            },
+        if 'organizer' not in body and booking.user.email:
+            body['organizer'] = {
+                'email': booking.user.email,
+            },
+        if 'recurrence' not in body and booking.recurrence:
+            f = booking.recurrence['frequency']
+            date_str = booking.recurrence['repeat_until'].replace('-', '')
+            body['recurrence'] = [
+                f'RRULE:FREQ=DAILY;INTERVAL={f};UNTIL={date_str}T235959Z',
+            ]
+
+        return body
+
+    def try_create_event(self, booking):
+        try:
+            return gcal_service.events().insert(
+                calendarId=self.calendar_id,
+                body=self._create_event_body(booking),
+            ).execute()
+
+        except HttpError as e:
+            logger.error(e)
+
+        return None
+
+    def try_update_event(self, booking):
+        try:
+            event = self.try_get_event(booking)
+
+            return gcal_service.events().update(
+                calendarId=self.calendar_id,
+                eventId=booking.google_calendar_event_id,
+                body=self._create_event_body(booking, event),
+            ).execute()
+
+        except HttpError as e:
+            logger.error(e)
+
+        return None
+
+    """
+    Get the Google Calendar event that corresponds to a certain booking.
+    """
+    def try_get_event(self, booking):
+        try:
+            event = gcal_service.events().get(
+                calendarId=self.calendar_id,
+                eventId=booking.google_calendar_event_id,
+            ).execute()
+            return event
+
+        except HttpError as e:
+            logger.error(e)
+
+        return None
+
+    '''
+    Delete the Google Calendar event for a booking, if it exist.
+    '''
+    def try_delete_event(self, booking):
+        try:
+            gcal_service.events().delete(
+                calendarId=self.calendar_id,
+                eventId=booking.google_calendar_event_id
+            ).execute()
+            return True
+
+        except HttpError as e:
+            error = json.loads(e.content)['error'] # = { errors, code, message }
+
+            # If the event has already been deleted, OK
+            if error['code'] == 410:
+                return True
+            logger.error(e)
+
+        return False
+
+
+class GCalCreateEventThread(Thread):
+    def __init__(self, booking):
+        self.booking = booking
+        self.gcal = GoogleCalendar(booking.bookable.google_calendar_id)
+        Thread.__init__(self)
+
+    def run(self):
+        if self.booking.google_calendar_event_id:
+            raise Exception(f'Booking already has a Google Calendar event')
+        event = self.gcal.try_create_event(self.booking)
+        if event:
+            self.booking.google_calendar_event_id = event['id']
+            self.booking.save(skip_gcal=True, update_fields=['google_calendar_event_id'])
```

Comment:
Oooh my... You have two copies of a stateful object in two separate threads, each running the same method.
To be fair, the thread is started after the main thread has saved to DB, and you do make sure to pass `skip_gcal=True` here to avoid loops, none the less this is a massive loaded shotgun pointing straight at your foot. I'd strongly advise to revise this implementation.

Reply:
So maybe a separate object containing only the GCal event <-> Booking relation? Or do you have any other suggestions?

Reply:
I would not make a deep copy of the model object, because I don't know all the state that django will track within the object. Instead I'd create a brand new object in the separate thread.
It's kind of a tricky situation since you need to update the model object twice, once to process the original request and again when the gcal sync is complete.
One option might be to track gcal events in a separate model, with a foreign key to bookings. This way there is no shared resource.

Reply:
Added a separate model for tracking the gcal events, as you suggested.